corehq/apps/reports/standard/cases/case_data.py

```diff
@@ -555,6 +561,255 @@ def close_case_view(request, domain, case_id):
+    """
+    @define
+    class DeleteCase:
+        name = field()
+        url = field()
+        is_primary = field(default=False)
+        delete_forms = field(factory=list)
+
+    @define
+    class DeleteForm:
+        name = field()
+        url = field()
+        affected_cases = field(factory=list)
+
+    @define
+    class FormAffectedCases:
+        case_name = field(default=None)
+        is_current_case = field(default=False)
+        actions = field(factory=list)
+
+    @define
+    class AffectedCase:
+        id = field()
+        name = field()
+        url = field()
+        affected_forms = field(factory=list)
+
+    def get_affected_case(case_id):
+        for affected_case in affected_cases_display:
+            if affected_case.id == case_id:
+                return affected_case
+        affected_case = AffectedCase(id=case_id, name=None, url=None)
+        affected_cases_display.append(affected_case)
+        return affected_case
+
+    @define
+    class AffectedForm:
+        name = field()
+        url = field()
+        actions = field()
+
+    @define
+    class ReopenedCase:
+        name = field()
+        url = field()
+        closing_form = field()
+
+    delete_cases = []  # list of cases to be soft deleted
+    delete_forms = []  # list of forms to be soft deleted
+
+    # For formatting the list of cases/submission forms
+    form_names = {}
+
+    delete_cases_display = []
+    reopened_cases_display = []
+    affected_cases_display = []
+
+    update_actions = [const.CASE_ACTION_INDEX,
+                      const.CASE_ACTION_UPDATE,
+                      const.CASE_ACTION_ATTACHMENT,
+                      const.CASE_ACTION_COMMTRACK,
+                      const.CASE_ACTION_REBUILD]
+
+    def walk_case_relations(case, subcase_count):
+        if case.case_id not in delete_cases:
+            delete_cases.append(case.case_id)
+            if len(delete_cases) > MAX_CASE_COUNT or subcase_count >= MAX_SUBCASE_DEPTH:
+                raise TooManyCases("Too many cases to delete")
+            current_case = DeleteCase(name=case.name, url=reverse('case_data', args=[domain, case.case_id]))
+            if len(delete_cases) == 1:  # only add primary label to the main case
+                current_case.is_primary = True
+            delete_cases_display.append(current_case)
+        xforms = case.xform_ids
+        case_xforms = []
+        for xform in xforms:
+            if xform not in case_xforms:
+                case_xforms.append(xform)
+
+        # iterating through all non-archived forms related to the case
+        for form_id in case_xforms:
+            if form_id not in delete_forms:
+                delete_forms.insert(0, form_id)
+            form_object = XFormInstance.objects.get_form(form_id, domain)
+            if form_id not in form_names:
+                form_names[form_id] = escape(xmlns_to_name(domain, form_object.xmlns, form_object.app_id))
```

Comment:
as an example guidance to further break down this function...
could all the places here where `form_names[form_id]` is being set be converted into a utility function that looks like...

```
def get_form_name(domain, form_object):
    form_name = escape(xmlns_to_name(domain, form_object.xmlns, form_object.app_id))
    if form_name == form_object.xmlns:
        inferred_name = [
            get_app(domain, form_object.app_id).name or "[Unknown App]",
            "[Unknown Module]",
             form_object.name or "[Unknown Form]"
         ]
         form_name = escape(' > '.join(inferred_name))
    return form_name
```
then right before you set `current_form`, you can do...
```
form_names[form_id] = get_form_name(domain, form_object)
```

Reply:
I ended up puling this and the process that gets the case's xforms into their own functions but no matter what I tried, I couldn't break down `walk_case_relations` further, partly because everything I tried ended up being very redundant and required passing around a lot of paramaters/context when it'd be easier to just access them in the main function. Maybe I'm thinking about this in the wrong way? 

I think even just pulling out all the dataclass definitions and those 2 functions make `get_case_and_display_data` as a whole, way more digestable. Do you have any more thoughts on this?